server/src/service-docs/extensions.ts

```diff
@@ -0,0 +1,73 @@
+import { buildMessage, ValidateBy, ValidationOptions } from 'class-validator';
+
+export type ExtensionKey = string;
+export type ExtensionValueType =
+  | ExtensionPrimitiveValueType
+  | ExtensionPrimitiveValueType[];
+export type ExtensionPrimitiveValueType = string | number | boolean;
+export type ExtensionObject = Record<ExtensionKey, ExtensionValueType>;
+
+export function isExtensionPrimitiveValueType(
+  extensionValue: unknown,
+): extensionValue is ExtensionPrimitiveValueType {
+  return (
+    typeof extensionValue === 'string' ||
+    typeof extensionValue === 'number' ||
+    typeof extensionValue === 'boolean'
+  );
+}
+
+export function isExtensionValueType(
+  extensionValue: unknown,
+): extensionValue is ExtensionValueType {
+  if (Array.isArray(extensionValue)) {
+    for (const item of extensionValue) {
+      if (!isExtensionPrimitiveValueType(item)) {
+        return false;
+      }
+    }
+  } else if (!isExtensionPrimitiveValueType(extensionValue)) {
+    return false;
+  }
+
+  return true;
```

Comment:
This is just an optional code style improvement - if you like your version better, then just ignore this :D 

```suggestion
  if (Array.isArray(extensionValue)) {
    for (const item of extensionValue) {
      if (!isExtensionPrimitiveValueType(item)) {
        return false;
      }
    }

    return true;
  }

  return isExtensionPrimitiveValueType(extensionValue);
```